File: Technical/3 ClassNotes/Java/src/practiceProblems/SmartTrafficAnalysis.java
package practiceProblems;

import java.util.Arrays;
import java.util.HashMap;

public class SmartTrafficAnalysis {

    public static void main(String[] args) {

        int traffic[] =
                {10,0,25,30,0,15,20,5};

        System.out.println(
                "Original Traffic Data:");

        printArray(traffic);


        //Move zeros
        moveZeros(traffic);

        System.out.println(
                "\nAfter Moving Zeros:");

        printArray(traffic);


        //Pair Sum
        int target = 35;

        System.out.println(
                "\nChecking Pair Sum:");

        pairSum(traffic,target);


        //Sliding Window
        int k = 3;

        int maxTraffic =
                maxVehiclesInWindow(
                        traffic,k);

        System.out.println(
                "\nMaximum vehicles in "
                + k
                + " consecutive minutes: "
                + maxTraffic);

    }



    //Print Array
    static void printArray(int arr[]) {

        for(int i : arr)
            System.out.print(i + " ");

        System.out.println();
    }



    //Move Zeros
    static void moveZeros(int arr[]) {

        int left = 0;

        for(int right=0;
            right<arr.length;
            right++) {

            if(arr[right] != 0) {

                int temp =
                        arr[left];

                arr[left] =
                        arr[right];

                arr[right] =
                        temp;

                left++;
            }
        }
    }



    //Pair Sum using HashMap
    static void pairSum(int arr[],
                        int target) {

        HashMap<Integer,Integer>
                map =
                new HashMap<>();

        for(int i : arr) {

            int diff =
                    target-i;

            if(map.containsKey(diff)) {

                System.out.println(
                        "Pair Found: "
                        + diff
                        + " "
                        + i);

                return;
            }

            map.put(i,1);
        }

        System.out.println(
                "No Pair Found");
    }



    //Fixed Sliding Window
    static int maxVehiclesInWindow(
            int arr[],
            int k) {

        int windowSum = 0;

        for(int i=0;i<k;i++)
            windowSum += arr[i];

        int maxSum = windowSum;

        for(int i=k;i<arr.length;i++) {

            windowSum =
                    windowSum
                    - arr[i-k]
                    + arr[i];

            maxSum =
                    Math.max(
                            maxSum,
                            windowSum);
        }

        return maxSum;
    }

}

File: Technical/3 ClassNotes/Python/practice_problems/SmartTrafficAnalysis.py
class SmartTrafficAnalysis:

    # Print Array
    def print_array(self, arr):

        for i in arr:
            print(i, end=" ")

        print()


    # Move Zeros
    def move_zeros(self, arr):

        left = 0

        for right in range(len(arr)):

            if arr[right] != 0:

                arr[left], arr[right] = \
                    arr[right], arr[left]

                left += 1


    # Pair Sum using HashMap(Dictionary)
    def pair_sum(self, arr, target):

        freq = {}

        for i in arr:

            diff = target - i

            if diff in freq:

                print(
                    "Pair Found:",
                    diff,
                    i
                )

                return

            freq[i] = 1

        print("No Pair Found")


    # Fixed Sliding Window
    def max_vehicles_in_window(
            self,
            arr,
            k):

        window_sum = sum(arr[:k])

        max_sum = window_sum

        for i in range(k, len(arr)):

            window_sum = \
                window_sum \
                - arr[i-k] \
                + arr[i]

            max_sum = max(
                max_sum,
                window_sum
            )

        return max_sum



# Driver Code

obj = SmartTrafficAnalysis()

traffic = [10,0,25,30,0,15,20,5]

print("Original Traffic Data:")

obj.print_array(traffic)


# Move Zeros
obj.move_zeros(traffic)

print("\nAfter Moving Zeros:")

obj.print_array(traffic)


# Pair Sum
target = 35

print("\nChecking Pair Sum:")

obj.pair_sum(traffic, target)


# Sliding Window
k = 3

max_traffic = \
    obj.max_vehicles_in_window(
        traffic,
        k
    )

print(
    "\nMaximum vehicles in",
    k,
    "consecutive minutes:",
    max_traffic
)